gbpservice/neutron/nsf/db/nsf_db.py

```diff
@@ -0,0 +1,405 @@
+            'ha_monitoring_data_network')
+        if not ha_monitoring_data_network:
+            nsd_db.ha_monitoring_data_network = None
+            return
+        with session.begin(subtransactions=True):
+            network_info_db = nsf_db_model.NetworkInfo(
+                id=ha_monitoring_data_network['id'],
+                network_policy=ha_monitoring_data_network['network_policy'])
+            session.add(network_info_db)
+            session.flush()
+            nsd_db.ha_monitoring_data_network = (
+                ha_monitoring_data_network['id'])
+            del network_service_device['ha_monitoring_data_network']
+
+    def create_network_service_device(self, session, network_service_device):
+        with session.begin(subtransactions=True):
+            network_service_device_db = nsf_db_model.NetworkServiceDevice(
+                id=(network_service_device.get('id')
+                    or uuidutils.generate_uuid()),
+                name=network_service_device['name'],
+                description=network_service_device.get('description'),
+                tenant_id=network_service_device['tenant_id'],
+                mgmt_ip_address=network_service_device[
+                    'mgmt_ip_address'],
+                service_vendor=network_service_device.get('service_vendor'),
+                max_interfaces=network_service_device['max_interfaces'],
+                reference_count=network_service_device['reference_count'],
+                interfaces_in_use=network_service_device['interfaces_in_use'],
+                status=network_service_device['status'])
+            session.add(network_service_device_db)
+            self._set_mgmt_ports_for_nsd(
+                session, network_service_device_db, network_service_device)
+            self._set_ha_monitoring_data_port_for_nsd(
+                session, network_service_device_db, network_service_device)
+            self._set_ha_monitoring_data_network_for_nsd(
+                session, network_service_device_db, network_service_device)
+            return self._make_network_service_device_dict(
+                network_service_device_db)
+
+    def _get_network_service_device(self, session, network_service_device_id):
+        try:
+            return self._get_by_id(
+                session,
+                nsf_db_model.NetworkServiceDevice,
+                network_service_device_id)
+        except exc.NoResultFound:
+            raise nsf_exc.NetworkServiceDeviceNotFound()
+
+    def update_network_service_device(self, session, network_service_device_id,
+                                      updated_network_service_device):
+        with session.begin(subtransactions=True):
+            network_service_device_db = self._get_network_service_device(
+                session, network_service_device_id)
+            if updated_network_service_device.get('mgmt_data_ports'):
+                self._set_mgmt_ports_for_nsd(
```

Comment:
For update, we clear the existing ones and set them again because we do not know which ones are modified